govcd/system.go

```diff
@@ -43,6 +58,134 @@ func CreateOrg(vcdClient *VCDClient, name string, fullName string, description s
 
 }
 
+// Returns the UUID part of an entity ID
+// From "urn:vcloud:vdc:72fefde7-4fed-45b8-a774-79b72c870325",
+// will return "72fefde7-4fed-45b8-a774-79b72c870325"
+// From "urn:vcloud:catalog:97384890-180c-4563-b9b7-0dc50a2430b0"
+// will return "97384890-180c-4563-b9b7-0dc50a2430b0"
+func getBareEntityId(entityId string) (string, error) {
+	// Regular expression to match an ID:
+	//     3 strings (alphanumeric + "-") separated by a colon (:)
+	//     1 group of 8 hexadecimal digits
+	//     3 groups of 4 hexadecimal
+	//     1 group of 12 hexadecimal digits
+	reGetID := regexp.MustCompile(`^[\w-]+:[\w-]+:[\w-]+:([a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12})$`)
+	matchList := reGetID.FindAllStringSubmatch(entityId, -1)
+
+	// matchList has the format
+	// [][]string{[]string{"TOTAL MATCHED STRING", "CAPTURED TEXT"}}
+	// such as
+	// [][]string{[]string{"urn:vcloud:catalog:97384890-180c-4563-b9b7-0dc50a2430b0", "97384890-180c-4563-b9b7-0dc50a2430b0"}}
+	if len(matchList) == 0 || len(matchList[0]) < 2 {
+		return "", fmt.Errorf("error extracting ID from '%s'", entityId)
+	}
+	return matchList[0][1], nil
+}
+
+// Creates an edge gateway using its parents entities
+// https://code.vmware.com/apis/442/vcloud-director/doc/doc/operations/POST-CreateEdgeGateway.html
+func CreateEdgeGateway(vcdClient *VCDClient, egwc EdgeGatewayCreation) (Task, error) {
```

Comment:
In addition to this I would probably vote for that "wider" function which accepts `types.EdgeGateway`. While it is more complicated (and is harder, if possible at all, to validate), it would allow more flexibility which may be needed in close future:

* One of them would be `HaEnabled` setting which should be used quite often in production environments. 
* The other I'm interested (for load balancers for now) is `< EdgeGatewayServiceConfiguration >` which [would allow to enable ](https://vdc-repo.vmware.com/vmwb-repository/dcr-public/76f491b4-679c-4e1e-8428-f813d668297a/a2555a1b-22f1-4cca-b481-2a98ab874022/doc/doc/types/NetworkServiceType.html) NatService, LoadBalancerService IpsecVpnService/GatewayIpsecVpnService, DhcpService/GatewayDhcpService, FirewallService, or StaticRoutingService. At least for load balancers I will have to use it to enable load balancing at edge gateway level.

API doc ref:
https://code.vmware.com/apis/72/vcloud/doc/doc/types/GatewayType.html

Reply:
The only practical problem with this approach is that to create an edge gateway we still need 5 parameters. 
There are three possibilities:

1. Additional function
`func CreateAndConfigureEdgeGateway(vcdClient *VCDClient, orgName, vdcName, egwName string, egwConfiguration *types.EdgeGateway ) (Task, error)` .  With this, the original creation function becomes a wrapper.

2. Add `egwConfiguration *types.EdgeGateway` to CreateEdgeGateway. If the configuration parameter is nil, it will be ignored.

3. Leave `CreateEdgeGateway` as it is, and add an `UpdateEdgeGateway` function that [updates the existing parameters](https://code.vmware.com/apis/553/vcloud-director/doc/doc/operations/PUT-EdgeGateway.html)

Reply:
> One of them would be HaEnabled setting which should be used quite often in production environments.

Actually, HaEnabled should be in the normal function and exposed through Terraform. It is there in the GUI after all.

Reply:
Done. The simplified structure has been updated, and a separate function that accepts a full structure has been added